File: Leetcode/动态规划_零钱兑换.py
"""
给你一个整数数组 coins ，表示不同面额的硬币；以及一个整数 amount ，表示总金额。

计算并返回可以凑成总金额所需的 最少的硬币个数 。如果没有任何一种硬币组合能组成总金额，返回 -1 。

你可以认为每种硬币的数量是无限的。

示例 1：

输入：coins = [1, 2, 5], amount = 11
输出：3 
解释：11 = 5 + 5 + 1

示例 2：

输入：coins = [2], amount = 3
输出：-1
"""
from typing import List

class Solution(object):
    def coinChange(self, coins: List[int], amount: int) -> int:

        # 初始化 dp 数组，dp[i] 表示凑出金额 i 所需的最小硬币数
        # 初始值设为 amount+1，代表无穷大（因为最多也只会用 amount 个 1 元）
        dp = [amount + 1] * (amount + 1)

        # 初始化0，因为凑出金额 0 所需硬币数为 0
        dp[0] = 0

        # 动态规划填表（尝试 金额i 与 硬币面值coin）
        for i in range(1, amount + 1):
            for coin in coins:
                if i >= coin:
                    dp[i] = min(dp[i], dp[i-coin] + 1)

        # 判断结果
        if dp[amount] == amount + 1:
            return -1
        else:
            return dp[amount]

if __name__ == '__main__':
    solution = Solution()

    coins = [1, 2, 5]
    amount = 11
    print(solution.coinChange(coins, amount))

    coins_2 = [2] 
    amount_2 = 3
    print(solution.coinChange(coins_2, amount_2))
    

File: Leetcode/动态规划_零钱兑换.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
// 【自顶向下】递归 + 备忘录
var coinChange = function(coins, amount) {
    const memo = new Array(amount + 1).fill(-1);

    const dp = (amount) => {
        if (amount === 0) return 0;
        if (amount < 0) return -1;
        // 命中子集：直接返回
        if (memo[amount] !== -1) return memo[amount];

        let res = Infinity;

        // 遍历每个硬币
        for (let coin of coins) {
            // 递归计算子问题
            const subProblem = dp(amount - coin);
            // 子问题无解，跳过
            if (subProblem === -1) continue;
            // 子问题有解，更新结果
            res = Math.min(res, subProblem + 1);
        }

        // 更新备忘录
        memo[amount] = (res === Infinity ? -1 : res);
        // 返回结果
        return memo[amount];
    };

    return dp(amount);
};

// 【自底向上】DP中的 for 循环 动态规划
var coinChange = function(coins, amount) {
    // dp数组，dp[i]表示 凑成 金额i 所需的 最少的硬币个数
    let dp = new Array(amount + 1).fill(amount + 1);

    de[0] = 0;

    // 循环每个 金额i
    for(let i = 1; i <= amount; i++) {
        // 依次从硬币类型 coins 中取出一个硬币
        for(let coin of coins){
            // 各个金额i >= 硬币coin,更新dp[i]
            if(i >= coin) {
                dp[i] = Math.min(dp[i], dp[i - coin] + 1);
            }
        }
    }
    return dp[amount] === amount + 1 ? -1 : dp[amount];
}
